pilot/pkg/dns/dns.go

```diff
@@ -312,6 +327,45 @@ func (h *LocalDNSServer) queryUpstream(upstreamClient *dns.Client, req *dns.Msg,
 	return response
 }
 
+func (u *upstreamServer) initConnection(uc *dns.Client) {
+	conn, err := uc.Dial(u.address)
+	if err != nil {
+		log.Warnf("unable to establish connection for upstream %s: %v", u.address, err)
+		return
+	}
+	u.Lock()
+	u.c = conn
+	u.Unlock()
+}
+
+// Inspired by from https://github.com/coredns/coredns/blob/fbf3f07f469a99fcbb5985a41c260a3fad26f908/plugin/forward/health.go#L76.
+func (u *upstreamServer) maybeResetConnection(upstreamClient *dns.Client) {
+	// It is OK to a function wide lock here and reset connection. This happens rarely.
+	u.Lock()
```

Comment:
One concern: say I send 100 concurrent DNS requests, they all fail so we hit this code. Now we will initiate 100 healthcheck+connection establishments? Seems we only need one

Reply:
Yes. I was thinking about that wanted to see if the general approach is fine - before make that change. I will think about it.

Reply:
Yeah the approach seems reasonable to me but I don't really know dns server best practices.

It seems like there are two possible options:
* On failure, async reconnect and use the new connection for new requests once done (this PR)
* On failure, reconnect and block all new requests until we have the new connection

First one is concerning since if we have a ""bad"" connection we will get a few failures until reconnect. Second one is concerning since we may see high latency and tie many requests success to a single connection establishment. I am not sure which is worse.

Reply:
Agree, concerned this will break down dns server